src/util/git.js

```diff
@@ -385,89 +367,82 @@ export default class Git {
-      this.supportsArchive = false;
+  async useDefaultBranchWithSymref(): Promise<void> {
+    const stdout = await Git.spawn(['ls-remote', '--symref', this.gitUrl.repository, 'HEAD']);
+    const lines = stdout.split('\n');
+    const [, ref] = lines[0].split(/\s+/);
+    const [sha] = lines[1].split(/\s+/);
+    this.ref = ref;
+    this.hash = sha;
+  }
 
-      if (!this.fetched) {
-        // in fact, `git archive` can't be used, and we haven't fetched the project yet. Do it now.
-        await this.fetch();
-      }
-      return (this.ref = this.hash = hash);
+  async useDefaultBranchWithLegacyGit(refs: GitRefs): Promise<void> {
+    const stdout = await Git.spawn(['ls-remote', this.gitUrl.repository, 'HEAD']);
+    const [sha] = stdout.split(/\s+/);
+    const resolvedResult = await resolveVersion(this.config, sha, refs);
+    if (resolvedResult !== false && !resolvedResult.defaultBranch) {
+      this.hash = resolvedResult.sha;
+      this.ref = resolvedResult.ref || '';
+    } else {
+      this.hash = sha;
+      this.ref = '';
     }
+  }
 
-    const ref = await this.findResolution(hash, names);
-    const commit = refs[ref];
-    if (commit) {
-      this.ref = ref;
-      return (this.hash = commit);
-    } else {
-      throw new MessageError(this.reporter.lang('couldntFindMatch', ref, names.join(','), this.gitUrl.repository));
+  async useDefaultBranch(refs: GitRefs): Promise<void> {
+    try {
+      await this.useDefaultBranchWithSymref();
+    } catch (err) {
+      // older versions of git don't support "--symref"
+      await this.useDefaultBranchWithLegacyGit(refs);
     }
   }
 
   /**
-   * Parse Git ref lines into hash of tag names to SHA hashes
+   * TODO description
    */
 
-  static parseRefs(stdout: string): GitRefs {
-    // store references
-    const refs = {};
-
-    // line delimited
-    const refLines = stdout.split('\n');
-
-    for (const line of refLines) {
-      const match = gitRefLineRegex.exec(line);
-
-      if (match) {
-        const [, sha, tagName] = match;
-
-        // As documented in gitrevisions:
-        //   https://www.kernel.org/pub/software/scm/git/docs/gitrevisions.html#_specifying_revisions
-        // "A suffix ^ followed by an empty brace pair means the object could be a tag,
-        //   and dereference the tag recursively until a non-tag object is found."
-        // In other words, the hash without ^{} is the hash of the tag,
-        //   and the hash with ^{} is the hash of the commit at which the tag was made.
-        const name = removeSuffix(tagName, '^{}');
+  async setRef(refs: GitRefs): Promise<string> {
+    // get commit ref
+    const {hash: version} = this;
 
-        refs[name] = sha;
-      }
+    const resolvedResult = await resolveVersion(this.config, version, refs);
+    if (resolvedResult === false) {
+      throw new MessageError(
+        this.reporter.lang('couldntFindMatch', version, Object.keys(refs).join(','), this.gitUrl.repository),
```

Comment:
This can result in some unreadable stuff like Couldn't find match for ">=1.5,<1.6" in "refs/heads/design_experimental,refs/heads/experimental_modal_scroll,refs/heads/master,refs/merge-requests/1/head,refs/merge-requests/10/head,refs/merge-requests/100/head,refs/merge-requests/101/head,refs/merge-requests/102/head,refs/merge-requests/103/head,refs/merge-requests/104/head,refs/merge-requests/105/head,refs/merge-requests/106/head,refs/merge-requests/107/head,refs/merge-requests/108/head,refs/merge-requests/11/head,refs/merge-requests/12/head,refs/merge-requests/13/head,refs/merge-requests/14/head,refs/merge-requests/15/head,refs/merge-requests/16/head,refs/merge-requests/17/head,refs/merge-requests/18/head,refs/merge-requests/19/head,refs/merge-requests/2/head,refs/merge-requests/20/head,refs/merge-requests/21/head,refs/merge-requests/22/head,refs/merge-requests/23/head,refs/merge-requests/24/head,refs/merge-requests/25/head,refs/merge-requests/26/head,refs/merge-requests/27/head,refs/merge-requests/28/head,refs/merge-requests/29/head,refs/merge-requests/3/head,refs/merge-requests/30/head,refs/merge-requests/31/head,refs/merge-requests/32/head,refs/merge-requests/33/head,refs/merge-requests/34/head,refs/merge-requests/35/head,refs/merge-requests/36/head,refs/merge-requests/37/head,refs/merge-requests/38/head,refs/merge-requests/39/head,refs/merge-requests/4/head,refs/merge-requests/40/head,refs/merge-requests/41/head,refs/merge-requests/42/head,refs/merge-requests/43/head,refs/merge-requests/44/head,refs/merge-requests/45/head,refs/merge-requests/46/head,refs/merge-requests/47/head,refs/merge-requests/48/head,refs/merge-requests/49/head,refs/merge-requests/5/head,refs/merge-requests/50/head,refs/merge-requests/51/head,refs/merge-requests/52/head,refs/merge-requests/53/head,refs/merge-requests/54/head,refs/merge-requests/55/head,refs/merge-requests/56/head,refs/merge-requests/57/head,refs/merge-requests/58/head,refs/merge-requests/59/head,refs/merge-requests/6/head,refs/merge-requests/60/head,refs/merge-requests/61/head,refs/merge-requests/62/head,refs/merge-requests/63/head,refs/merge-requests/64/head,refs/merge-requests/65/head,refs/merge-requests/66/head,refs/merge-requests/67/head,refs/merge-requests/68/head,refs/merge-requests/69/head,refs/merge-requests/7/head,refs/merge-requests/70/head,refs/merge-requests/71/head,refs/merge-requests/72/head,refs/merge-requests/73/head,refs/merge-requests/74/head,refs/merge-requests/75/head,refs/merge-requests/76/head,refs/merge-requests/77/head,refs/merge-requests/78/head,refs/merge-requests/79/head,refs/merge-requests/8/head,refs/merge-requests/80/head,refs/merge-requests/81/head,refs/merge-requests/82/head,refs/merge-requests/83/head,refs/merge-requests/84/head,refs/merge-requests/85/head,refs/merge-requests/86/head,refs/merge-requests/87/head,refs/merge-requests/88/head,refs/merge-requests/89/head,refs/merge-requests/9/head,refs/merge-requests/90/head,refs/merge-requests/91/head,refs/merge-requests/92/head,refs/merge-requests/93/head,refs/merge-requests/94/head,refs/merge-requests/95/head,refs/merge-requests/96/head,refs/merge-requests/97/head,refs/merge-requests/98/head,refs/merge-requests/99/head,refs/tags/1.0.0,refs/tags/1.0.1,refs/tags/1.0.10,refs/tags/1.0.11,refs/tags/1.0.3,refs/tags/1.0.4,refs/tags/1.0.5,refs/tags/1.0.6,refs/tags/1.0.7,refs/tags/1.0.8,refs/tags/1.0.9,refs/tags/1.1.0,refs/tags/1.1.1,refs/tags/1.1.2,refs/tags/1.1.3,refs/tags/1.1.4,refs/tags/1.2.0,refs/tags/1.2.1,refs/tags/1.2.2,refs/tags/1.3.0,refs/tags/1.4.0,refs/tags/1.4.1,refs/tags/1.4.2,refs/tags/1.4.3,refs/tags/1.4.4,refs/tags/1.4.5,refs/tags/1.4.6,refs/tags/1.4.7,refs/tags/1.5.0,refs/tags/1.5.1,refs/tags/1.6.0,refs/tags/1.6.1,refs/tags/1.6.2,refs/tags/1.7.0,refs/tags/1.7.1" for "https://gitlab.com/leanlabsio/kanban.git"